### PR TITLE
Load images in CanvasKit with Cookies

### LINKED ISSUE #150351: [web][CanvasKit][skwasm] Cookie-based authentication for Network images not possible
### Steps to reproduce

1. Download the example project:
[image_cookies.zip](https://github.com/user-attachments/files/15871460/image_cookies.zip)
2. Generate a trusted SSL certificate for localhost
3. Run the server
4. Run the app

### Expected results

The `/authenticated_image` request succeeds as it can validate the passed Cookies.

### Actual results

The cookies are not sent as part of the `/authenticated_image` request, and thus, authenticating failed.

### Code sample

Full runnable example: [image_cookies.zip](https://github.com/user-attachments/files/15871460/image_cookies.zip)

<details open><summary>Code sample</summary>

`main.dart`:
```dart
import 'package:dio/browser.dart';
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Web Cookie Authentication'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool _loaded = false;

  @override
  void initState() {
    super.initState();
    _getAuthCookies();
  }

  void _getAuthCookies() async {
    final dio = Dio();
    var adapter = BrowserHttpClientAdapter();
    adapter.withCredentials = true;
    dio.httpClientAdapter = adapter;
    await dio.getUri(Uri.parse('https://localhost:3000'));
    if (!mounted) return;
    setState(() {
      _loaded = true;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            if (!_loaded)
              const Text(
                'Loading',
              ),
            if (_loaded)
              Image.network('https://localhost:3000/authenticated_image')
          ],
        ),
      ),
    );
  }
}
```

`server.ts`:
```ts
import express from 'express';
import cors from 'cors';
import * as fs from 'fs';
import cookieParser from "cookie-parser";
import * as https from 'node:https';

const key = fs.readFileSync('./cert/localhost/localhost.decrypted.key');
const cert = fs.readFileSync('./cert/localhost/localhost.crt');

const app = express();

app.use(
  cors({
    origin: /http:\/\/localhost:\d{4,5}/,
    credentials: true,
  })
);
app.use(cookieParser())

app.get('/', (req, res) => {
  res.cookie('Auth-Cookie', "valid_request", {
    sameSite: 'none',
    maxAge: 1000 * 60 * 60,
    secure: true,
  });
  res.send('<img src="/authenticated_image">');
});

app.get('/authenticated_image', (req, res) => {
  const authCookie = req.cookies['Auth-Cookie'];
  if (authCookie === 'valid_request') {
    fs.createReadStream('./src/favicon.png').pipe(res);
  } else {
    res.status(403).send('Unauthenticated');
  }
});

const server = https.createServer({key, cert}, app);
const port = 3000;
server.listen(port, () => {
  console.log(`Server is listening on https://localhost:${port}`);
});
```
</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[!] Flutter (Channel [user-branch], 3.23.0-13.0.pre.243, on macOS 14.3 23D56 darwin-arm64, locale en-DE)
    ! Flutter version 3.23.0-13.0.pre.243 on channel [user-branch] at /Users/dennis/develop/bin/flutter
      Currently on an unknown channel. Run `flutter channel` to switch to an official channel.
      If that doesn't fix the issue, reinstall Flutter by following instructions at https://flutter.dev/docs/get-started/install.
    ! Upstream repository https://github.com/simpleclub-extended/flutter.git is not a standard remote.
      Set environment variable "FLUTTER_GIT_URL" to https://github.com/simpleclub-extended/flutter.git to dismiss this error.
[!] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.
[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.2)
[✓] IntelliJ IDEA Ultimate Edition (version 2024.1)
[✓] VS Code (version 1.88.1)
[✓] Connected device (3 available)
[✓] Network resources

! Doctor found issues in 2 categories.
```

Note: I'm on a custom branch as I've experimented with a fix.
At the time of testing, no changes were made to Flutter. :)

</details>

### PR DESCRIPTION
**Pull request by IchordeDionysos**
Typically, image requests in a web environment are sent with Cookies.

In CanvasKit, images are not loaded with cookies, making it impossible to use cookie-based authentication for image requests.

Fixes #150351

> ⚠️ Please review this carefully to make sure this change does not to introduce any vulnerability. :)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by IchordeDionysos** [comment]
The test failures are not related to my changes?!

**Comment by IchordeDionysos** [comment]
Damn @goderbauer shouldn't have merged main...
It removed the web label again 😢

File: packages/flutter/test/painting/_network_image_test_web.dart [hunk 1]
```diff
@@ -92,4 +92,27 @@ void runTests() {
     expect(tester.takeException().toString(),
         'HTTP request failed, statusCode: 200, https://www.example.com/images/frame3.png');
   });
+
+  testWidgets('loads an image using withCredentials true',
+      (WidgetTester tester) async {
+    if (!isCanvasKit) {
+      // The HTML renderer uses an `<img>` tag, it does not fetch the image via XMLHttpRequest.
+      return;
```
Comment (kevmoo) [inline packages/flutter/test/painting/_network_image_test_web.dart:100]: Would it be better to use https://pub.dev/documentation/test/latest/index.html#skipping-tests ??

Comment (kevmoo) [reply packages/flutter/test/painting/_network_image_test_web.dart:100]: Or maybe not. It adds noise to the output...

**Comment by kevmoo** [comment]
Thoughts @yjbanov ?

**Comment by yjbanov** [comment]
> Thoughts @yjbanov ?

This is something worth looking into, but to review this properly it requires proper understanding of how `withCredentials` works. In particular, this PR sets `withCredentials` on all network image request, which can alter the current behavior. I don't know enough to tell if it could be a breaking change. Unfortunately, I'm dealing with other stuff, so I'm unable to review this yet. If anyone is available to do a proper review, do feel free to take over.

**Comment by yjbanov** [comment]
My current inclination is that we should not merge this as is, because setting `withCredentials` on behalf of all Flutter Web apps is a major behavioral change. I can see a couple of alternatives:

- Make this parameter controllable by the user of `Image.network`, perhaps an additional constructor argument. However, I'm not sure what it would mean on mobile.
- Add this kind of customizability to https://pub.dev/packages/flutter_image instead of the core framework.

**Comment by goderbauer** [comment]
(triage) @IchordeDionysos Do you still have plans for this PR based on the feedback given above?

**Comment by Piinks** [comment]
Thank you for your contribution. I'm going to close this PR for now since there are outstanding comments, just to get this off our PR review queue. Please don't hesitate to re-open or submit a new PR if you have the time to address the review comments. Thanks!

**Pull request closed without merging**